### PR TITLE
[core][distributed] add ep group and all2all interface

### PR DESCRIPTION
**Pull request by youkaichao**
the all2all interface should be refactored with a follow-up from https://github.com/vllm-project/vllm/pull/15956 .

this PR mainly adds ep (expert parallel) group, and all2all base class, so that backend-dependent code like pplx's `pplx_init` will not invade into common code path (`vllm/distributed/parallel_state.py`) .

### CODE REVIEW

File: vllm/distributed/device_communicators/all2all.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# SPDX-License-Identifier: Apache-2.0
+import torch
+
+from vllm.forward_context import get_forward_context
+
+
+class All2AllBase:
+
+    def __init__(self, cpu_group, model):
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/all2all.py:9]: for how to use `cpu_group` to initialize pplx, see https://github.com/ppl-ai/pplx-kernels/issues/18 as an example.

here we have access to the model instance, so we can also get moe configs here, and we can assume all moe layers have the same config, without `_all_to_all_cache` .

Comment (youkaichao) [reply vllm/distributed/device_communicators/all2all.py:9]: from @nandor , 

```python
uid = nvshmem_get_unique_id() if rank == 0 else nvshmem_alloc_empty_unique_id()
torch.distributed.broadcast(uid, src=0)
nvshmem_init(uid, rank, world_size)
```

is not necessary for intranode code.

we can tell if we are in single node with https://github.com/vllm-project/vllm/blob/79a1d25bbd09420ad7d8631671a70cbcaf496cfb/vllm/distributed/parallel_state.py#L1127

**Comment by varun-sundar-rabindranath** [comment]
Thanks @youkaichao  - The abstractions generally look good . But the dispatch-combine calls for pplx input a very different set of information than the naive all2all implementation. I think we can cross that bridge when we add the pplx All2All implementation to the abstraction - But just adding it as a note.

**Comment by youkaichao** [comment]
> But the dispatch-combine calls for pplx input a very different set of information than the naive all2all implementation.

agree, right now the interface is just porting code from the naive implementation. we should definitely change it when we add the pplx All2All implementation to the abstraction.

File: vllm/distributed/device_communicators/all2all.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        # intra-node and inter-node communication
+        self.intranode = in_the_same_node_as(cpu_group, source_rank=0)
+        self.internode = not self.intranode
+
+    def dispatch(self, hidden_states: torch.Tensor,
+                 router_logits: torch.Tensor):
+        raise NotImplementedError
+
+    def combine(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/all2all.py:35]: Is there a function signature for dispatch and combine that will work across all kernels?
* The naive multicast implementation
* pplx-kernels
* DeepEP

I think it would to look like:
```
def dispatch(self, 
             hidden_states: torch.Tensor,
             hs_scales: torch.Tensor, # Quantized scales for hidden_states
             router_logits: torch.Tensor,  
             topk_weights: torch.Tensor,  
             topk_ids: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor] # out_hidden_states and out_scales
```
(There are other parameters to the pplx-kernels that are needed but I think they can be encapsulated in a PPLXAll2All wrapper class)

@varun-sundar-rabindranath @bnellnm WDYT?

Comment (tlrmchlsmth) [reply vllm/distributed/device_communicators/all2all.py:35]: Another consideration here is that the output format of the different A2A implementations will be different.

* For naive All2All, the output shape would be `[total_tokens_across_dp, hidden_size]`
* For pplx All2All, it will be `[num_experts_per_rank, max_tokens_per_expert, hidden_size]`, with padding along axis 1
* For DeepEP I am not sure what the output format will look like

Comment (youkaichao) [reply vllm/distributed/device_communicators/all2all.py:35]: for a generic dispatch/combine interface for different backends (pplx, naive, deepep), I think we can have:
```python
def dispatch(self, tensors: List[torch.Tensor]) -> List[torch.Tensor]:
def combine(self, tensors: List[torch.Tensor]) -> List[torch.Tensor]:
```
the backend can decide what does each tensor mean, and the moe layer can use assertion to make sure the list contains data it needs.

Comment (tlrmchlsmth) [reply vllm/distributed/device_communicators/all2all.py:35]: What's the reason for having the base class in this case? It seems that the caller will have to know exactly the implementation it's using

Comment (youkaichao) [reply vllm/distributed/device_communicators/all2all.py:35]: It's mainly to manage the lifecycle of the communication library, and separate the implementations.

it is true that the dispatch / combine apis are quite different, but what matters in this pr is to unify the construction, backend selection, and deconstruction. as a byproduct, we need to have a generic dispatch / combine function.

Comment (bnellnm) [reply vllm/distributed/device_communicators/all2all.py:35]: I agree with @tlrmchlsmth that it would be good if these interfaces matched the pplx AllToAll interfaces.

File: vllm/distributed/utils.py [hunk 3]
```diff
@@ -361,6 +361,11 @@ def stateless_destroy_torch_distributed_process_group(
...
     # Lazy import for non-CUDA backends.
-    from torch.distributed.distributed_c10d import _shutdown_backend
-    _shutdown_backend(pg)
+    try:
+        # pytorch <= 2.6
+        from torch.distributed.distributed_c10d import _shutdown_backend
+        _shutdown_backend(pg)
+    except ImportError:
+        # pytorch >= 2.7
+        pg.shutdown()
```
Comment (bnellnm) [inline vllm/distributed/utils.py:370]: I used this in the PPLX PR.  I think this is a bit nicer than try/catch.
```
    if is_torch_equal_or_newer("2.7"):
        pg.shutdown()
    else:
        # Lazy import for non-CUDA backends.                                                                                                                            
        from torch.distributed.distributed_c10d import _shutdown_backend
        _shutdown_backend(pg)
```

**Review by tlrmchlsmth — APPROVED** [review]

File: vllm/distributed/utils.py (hunk 3 above)
Comment (youkaichao) [reply vllm/distributed/utils.py:370]: feel free to change it in the moe modularization pr!

File: vllm/distributed/device_communicators/all2all.py (hunk 2 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/all2all.py:35]: feel free to change the interface in the moe modularization pr!
